In [1]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Loading datasets
train = pd.read_csv('train.csv')
campaign_data = pd.read_csv('campaign_data.csv')
coupon_item_mapping = pd.read_csv('coupon_item_mapping.csv')
customer_demographics = pd.read_csv('customer_demographics.csv')
customer_transaction_data = pd.read_csv('customer_transaction_data.csv')
item_data = pd.read_csv('item_data.csv')
test = pd.read_csv('test.csv')

Performing Data Pre-Processing and Data Cleaning in each DataFrame

In [4]:
#considering train dataframe
train.sample(10)

id  campaign_id  coupon_id  customer_id  redemption_status
14851   24480            8        433          416                  0
9288    15341           13        166           68                  0
72792  119552            8         13         1001                  0
48537   79575            8        983         1516                  0
56132   92309           13        886          426                  0
62456  102674           30        678          711                  0
74749  122730            8          8          775                  0
2997     4984            8        915          401                  0
70011  115011           13       1033          337                  0
51095   83820            8        700         1410                  0

In [5]:
train.shape

(78369, 5)

In [6]:
train.dtypes

id                   int64
campaign_id          int64
coupon_id            int64
customer_id          int64
redemption_status    int64
dtype: object

All the ids should be of type category,

redemption_status represents 2 categories : 0 for Coupon not redeemed, 1 - Coupon redeemed, it should be of type category

In [8]:
train['campaign_id'] = train['campaign_id'].astype('category')
train['coupon_id'] = train['coupon_id'].astype('category')
train['customer_id'] = train['customer_id'].astype('category')
train['id'] = train['id'].astype('category')
train['redemption_status'] = train['redemption_status'].astype('category')

In [9]:
#to verify the change
train.dtypes

id                   category
campaign_id          category
coupon_id            category
customer_id          category
redemption_status    category
dtype: object

In [10]:
# checking for missing values
train.isnull().sum()

id                   0
campaign_id          0
coupon_id            0
customer_id          0
redemption_status    0
dtype: int64

There are no missing values in the train dataframe

In [12]:
#checking for duplicated rows
train.duplicated().sum()

0

There are no duplicate rows in the train dataframe

In [14]:
# considering test dataframe
test.sample(10)

id  campaign_id  coupon_id  customer_id
41845  107071           18        543          407
24520   62720           20        721           33
14234   36464           23        266          566
3587     8921           20        389         1537
24799   63420           22        927          776
30619   78551           18       1014         1361
128       327           22        432          123
25900   66231           22        957         1496
15610   39981           18        898          922
17072   43684           22        432          817

In [15]:
test.dtypes

id             int64
campaign_id    int64
coupon_id      int64
customer_id    int64
dtype: object

All the ids should be converted to category

In [17]:
test['id'] = test['id'].astype('category')
test['campaign_id'] = test['campaign_id'].astype('category')
test['coupon_id'] = test['coupon_id'].astype('category')
test['customer_id'] = test['customer_id'].astype('category')

In [18]:
test.dtypes

id             category
campaign_id    category
coupon_id      category
customer_id    category
dtype: object

In [19]:
test.isnull().sum()

id             0
campaign_id    0
coupon_id      0
customer_id    0
dtype: int64

In [20]:
test.duplicated().sum()

0

Analysing campaign_data

In [22]:
# considering campaign_data dataframe
campaign_data.sample(10)

campaign_id campaign_type start_date  end_date
19            5             Y   12/01/13  15/02/13
10           13             X   19/05/13  05/07/13
13           10             Y   08/04/13  10/05/13
0            24             Y   21/10/13  20/12/13
7            19             Y   26/08/13  27/09/13
11           11             Y   22/04/13  07/06/13
6            18             X   10/08/13  04/10/13
27           26             X   12/08/12  21/09/12
14            9             Y   11/03/13  12/04/13
26           27             Y   25/08/12  27/10/12

In [23]:
campaign_data.shape

(28, 4)

In [24]:
campaign_data.dtypes

campaign_id       int64
campaign_type    object
start_date       object
end_date         object
dtype: object

campaign_type represents Anonymised Campaign Type (X/Y), it should be of type category.
start_date and end_date should be coverted to type datetime
campaign_id should be of type category

In [26]:
campaign_data['campaign_id'] = campaign_data['campaign_id'].astype('category')
campaign_data['campaign_type'] = campaign_data['campaign_type'].astype('category')
campaign_data['start_date'] = pd.to_datetime(campaign_data['start_date'], format='%d/%m/%y')
campaign_data['end_date'] = pd.to_datetime(campaign_data['end_date'], format='%d/%m/%y')


In [27]:
campaign_data.head()

campaign_id campaign_type start_date   end_date
0          24             Y 2013-10-21 2013-12-20
1          25             Y 2013-10-21 2013-11-22
2          20             Y 2013-09-07 2013-11-16
3          23             Y 2013-10-08 2013-11-15
4          21             Y 2013-09-16 2013-10-18

In [28]:
campaign_data.dtypes

campaign_id            category
campaign_type          category
start_date       datetime64[ns]
end_date         datetime64[ns]
dtype: object

In [29]:
#get a summary of numerical features of the dataset
campaign_data.describe()

start_date                       end_date
count                   28                             28
mean   2013-04-03 12:00:00  2013-05-15 08:34:17.142857216
min    2012-08-12 00:00:00            2012-09-21 00:00:00
25%    2012-12-20 18:00:00            2013-02-02 18:00:00
50%    2013-03-25 00:00:00            2013-04-26 00:00:00
75%    2013-08-14 00:00:00            2013-09-28 18:00:00
max    2013-10-21 00:00:00            2013-12-20 00:00:00

In [30]:
# statistical summary of categorical columns
campaign_data.describe(include="category")

campaign_id campaign_type
count            28            28
unique           28             2
top               1             Y
freq              1            22

From the above analysis, we observe that -:
1. There are 28 rows in campaign_data.
2. In campaign_type feature, Y is the most frequently occuring feature (22 out of 28) and N occurs 6 out of 28 times.
3. There are 28 unique campaign_ids identifying each row uniquely.
4. There no null values in coupon_item_mapping as the number of rows is same as number of count of both coulmns.

In [32]:
campaign_data.duplicated().sum()

0

In [33]:
#getting more information from start_date and end_date
campaign_data['campaign_duration'] = (campaign_data['end_date'] - campaign_data['start_date']).dt.days


campaign_duration represents no.of days the campaign happened

In [35]:
campaign_data.head()

campaign_id campaign_type start_date   end_date  campaign_duration
0          24             Y 2013-10-21 2013-12-20                 60
1          25             Y 2013-10-21 2013-11-22                 32
2          20             Y 2013-09-07 2013-11-16                 70
3          23             Y 2013-10-08 2013-11-15                 38
4          21             Y 2013-09-16 2013-10-18                 32

In [36]:
campaign_data.dtypes

campaign_id                category
campaign_type              category
start_date           datetime64[ns]
end_date             datetime64[ns]
campaign_duration             int64
dtype: object

In [37]:
#checking if campaign_duration is not zero neither less than 0 for any row
invalid_duration = campaign_data[campaign_data['campaign_duration'] <= 0]
invalid_duration

Empty DataFrame
Columns: [campaign_id, campaign_type, start_date, end_date, campaign_duration]
Index: []

Thus, no such row exists where campaign_duration is either 0 or less than 0

In [39]:
#creating sets for campaign_id for both the trian and campaign_data dataframe
train_set = set(train.campaign_id.values)
campaign_data_set = set(campaign_data.campaign_id.values)

In [213]:
#finding the campaign_id not in train_data dataframe
campaign_data_set - campaign_data_set.intersection(train_set)

{16, 17, 18, 19, 20, 21, 22, 23, 24, 25}

This shows that there 10 campaign_ids which are present in campaign_data  but not in train_data

Merging train, test data with campaign_data

In [43]:
train = pd.merge(train,campaign_data, on='campaign_id', how='left')
test = pd.merge(test,campaign_data, on='campaign_id', how='left')

In [44]:
train.shape

(78369, 9)

In [45]:
train.head()

id  campaign_id coupon_id customer_id redemption_status campaign_type  \
0  1           13        27        1053                 0             X   
1  2           13       116          48                 0             X   
2  6            9       635         205                 0             Y   
3  7           13       644        1050                 0             X   
4  9            8      1017        1489                 0             X   

  start_date   end_date  campaign_duration  
0 2013-05-19 2013-07-05                 47  
1 2013-05-19 2013-07-05                 47  
2 2013-03-11 2013-04-12                 32  
3 2013-05-19 2013-07-05                 47  
4 2013-02-16 2013-04-05                 48

In [46]:
test.head()

id  campaign_id coupon_id customer_id campaign_type start_date   end_date  \
0   3           22       869         967             X 2013-09-16 2013-10-18   
1   4           20       389        1566             Y 2013-09-07 2013-11-16   
2   5           22       981         510             X 2013-09-16 2013-10-18   
3   8           25      1069         361             Y 2013-10-21 2013-11-22   
4  10           17       498         811             Y 2013-07-29 2013-08-30   

   campaign_duration  
0                 32  
1                 70  
2                 32  
3                 32  
4                 32

Analysis of coupon_item_mapping

In [48]:
#considering coupon_item_mapping dataframe
coupon_item_mapping.sample(10)

coupon_id  item_id
77751         33    47083
15291         24    73088
66919         31    30663
58001         23    64750
31992       1068     6827
79938         31    16578
74668         32    22576
86267         29    61989
15952        960     9445
10483         37    57099

In [49]:
coupon_item_mapping.shape

(92663, 2)

In [50]:
coupon_item_mapping.dtypes

coupon_id    int64
item_id      int64
dtype: object

Both the ids are of type int, should be converted to category

In [52]:
coupon_item_mapping['coupon_id'] = coupon_item_mapping['coupon_id'].astype('category')
coupon_item_mapping['item_id'] = coupon_item_mapping['item_id'].astype('category')

In [53]:
coupon_item_mapping.dtypes

coupon_id    category
item_id      category
dtype: object

In [54]:
# statistical summary of categorical columns
coupon_item_mapping.describe(include="category")

coupon_id  item_id
count       92663    92663
unique       1116    36289
top            32    56523
freq        11814       11

From the above , we observe that:
1. We have 1,116 different coupons which are linked to 36,289 different items.
2. Among coupon_ids , 32 id occurs more frequently.
3. Among item_ids, 56523 id occurs more frequently.
4. There no null values in coupon_item_mapping as the number of rows is same as number of count of both ids.

In [56]:
coupon_item_mapping.duplicated().sum()

0

Analysis of customer_demographics

In [58]:
#considering customer_demographics dataframe
customer_demographics.sample(10)

customer_id age_range marital_status  rented family_size no_of_children  \
207          421     56-70        Married       0           2            NaN   
135          268       70+        Married       0           2            NaN   
47            89     46-55            NaN       0           1            NaN   
718         1483     46-55            NaN       0           2            NaN   
711         1471     46-55            NaN       0           1            NaN   
506         1051     46-55            NaN       0           2            NaN   
571         1190     56-70         Single       0           1            NaN   
633         1321     36-45        Married       0           3              1   
168          338     46-55            NaN       0           1            NaN   
279          557     26-35            NaN       0           3              2   

     income_bracket  
207               5  
135               2  
47                3  
718               4  
711               5  
506               2  
571               6  
633               1  
168               5  
279               4

In [59]:
customer_demographics.shape

(760, 7)

In [60]:
customer_demographics.dtypes

customer_id        int64
age_range         object
marital_status    object
rented             int64
family_size       object
no_of_children    object
income_bracket     int64
dtype: object

customer_id should be of type category
age-range represents age range of customer family in years , should be converted to category 
marital_status contains two categories: single and married, should be converted to category
rented represents 0 - not rented accommodation, 1 - rented accommodation, should be converted to category
income_bracket represents label encoded income bracket, it should be converted to ordinal category

In [62]:
#converting datatypes
customer_demographics['customer_id'] = customer_demographics['customer_id'].astype('category')
customer_demographics['age_range'] = customer_demographics['age_range'].astype('category')
customer_demographics['marital_status'] = customer_demographics['marital_status'].astype('category')
customer_demographics['rented'] = customer_demographics['rented'].astype('category')


In [63]:
customer_demographics['family_size'].value_counts()

family_size
2     303
1     248
3     104
5+     57
4      48
Name: count, dtype: int64

As the family_size contains mixed datatype, integers and string, converting it into category

In [65]:
customer_demographics['family_size'] = customer_demographics['family_size'].astype('category')

In [66]:
customer_demographics['no_of_children'].value_counts()

no_of_children
1     107
3+     60
2      55
Name: count, dtype: int64

Same as , no_of_children contains mixed datatype, integers and string, converting it into category

In [68]:
customer_demographics['no_of_children'] = customer_demographics['no_of_children'].astype('category')

In [69]:
customer_demographics['income_bracket'].value_counts()

income_bracket
5     187
4     165
6      88
3      70
2      68
1      59
8      37
7      32
9      29
12     10
10     10
11      5
Name: count, dtype: int64

In [70]:
#converting income_bracket into into ordinal ctegory
values = [1, 2, 3, 4, 5, 6, 7, 8, 9,10, 11, 12]
customer_demographics['income_bracket'] = pd.Categorical(
    customer_demographics['income_bracket'],
    categories=values,
    ordered=True
)


In [71]:
customer_demographics.sample(10)

customer_id age_range marital_status rented family_size no_of_children  \
138         276     46-55            NaN      0           2            NaN   
427         876     36-45         Single      0           1            NaN   
74          143       70+            NaN      0           1            NaN   
387         789     46-55            NaN      0           2            NaN   
600        1250     36-45        Married      0          5+             3+   
393         799       70+            NaN      0           1            NaN   
314         628     46-55        Married      0           3              1   
261         524     26-35            NaN      0           1            NaN   
28           52     36-45        Married      0          5+             3+   
496        1021     26-35            NaN      0           1            NaN   

    income_bracket  
138              5  
427              9  
74               2  
387              1  
600              8  
393              5  
314              4  
261              5  
28               7  
496              4

In [72]:
customer_demographics.dtypes

customer_id       category
age_range         category
marital_status    category
rented            category
family_size       category
no_of_children    category
income_bracket    category
dtype: object

In [73]:
#checking for missing values as we can see from above they are present in marital_status and  no_of_children
customer_demographics.isnull().sum()

customer_id         0
age_range           0
marital_status    329
rented              0
family_size         0
no_of_children    538
income_bracket      0
dtype: int64

In [74]:
# checking the missing value percentage
(customer_demographics.isnull().sum()/len(customer_demographics)) * 100

customer_id        0.000000
age_range          0.000000
marital_status    43.289474
rented             0.000000
family_size        0.000000
no_of_children    70.789474
income_bracket     0.000000
dtype: float64

Here, 43% of data in marital_status and around 71% data in no_of_children are missing, which is very huge amount

We will treat the missing values after merging the data

In [77]:
# #adding Unknown category to marital_status
# customer_demographics['marital_status'] = customer_demographics['marital_status'].cat.add_categories(["Unknown"])
# customer_demographics.fillna({'marital_status': "Unknown"}, inplace=True)

# #adding Unknown category to no_of_children
# customer_demographics['no_of_children'] = customer_demographics['no_of_children'].cat.add_categories(["Unknown"])
# customer_demographics.fillna({'no_of_children': "Unknown"}, inplace=True)



In [78]:
# customer_demographics['marital_status'].isnull().sum()

In [79]:
# customer_demographics['no_of_children'].isnull().sum()

In [80]:
#checking for duplicate rows
customer_demographics.duplicated().sum()

0

In [81]:
# statistical summary of categorical columns
customer_demographics.describe(include="category")

customer_id age_range marital_status  rented family_size  \
count           760       760            431     760         760   
unique          760         6              2       2           5   
top               1     46-55        Married       0           2   
freq              1       271            317     719         303   

       no_of_children  income_bracket  
count             222             760  
unique              3              12  
top                 1               5  
freq              107             187

In [82]:
customer_demographics['income_bracket'].value_counts()

income_bracket
5     187
4     165
6      88
3      70
2      68
1      59
8      37
7      32
9      29
10     10
12     10
11      5
Name: count, dtype: int64

From the above, we observe:
1. There are 760 unique values in customer_id .
2. In age_range, 6 unique ranges are present out of which 46-55 is the most common range in family. 
3. There are null values present in marital_status and no_of_children .
4. In family_size, there are 5 unique categories among which 2 is the most common family_size.
5. Thre are 12 unique categories in income_bracket, out of which 5 is the most common, which shows that people with very low income (1,2, 3 ) are less , middle range income people (4,5,6) has more data points and very high income range (7,8,9,10,11,12) are very few.


Analysis of customer_transaction_data

In [85]:
#considering customer_transaction_data dataframe
customer_transaction_data.sample(10)

date  customer_id  item_id  quantity  selling_price  \
163973   2012-05-05          746    48910         1         105.08   
586651   2012-10-10          464     6028         1          52.01   
1044724  2013-03-21          772    18501         1         102.94   
566361   2012-10-02         1367    19568         1          44.52   
271343   2012-06-16          918    12403         1          89.05   
785504   2012-12-19          244    14172         1          35.62   
1315970  2013-06-30         1442     5315         2         142.48   
232958   2012-06-01         1464    61434         1          27.07   
1117538  2013-04-17         1022    25718         1         284.60   
483841   2012-09-03         1475     6958         1          31.35   

         other_discount  coupon_discount  
163973             0.00              0.0  
586651            -5.70              0.0  
1044724            0.00              0.0  
566361           -26.36              0.0  
271343           -53.07              0.0  
785504           -22.08              0.0  
1315970          -17.81              0.0  
232958             0.00              0.0  
1117538            0.00              0.0  
483841           -25.29              0.0

In [86]:
customer_transaction_data.shape

(1324566, 7)

In [87]:
#checking for datatypes of each column
customer_transaction_data.dtypes

date                object
customer_id          int64
item_id              int64
quantity             int64
selling_price      float64
other_discount     float64
coupon_discount    float64
dtype: object

date should be coverted to datetime datatype and customer_id and item_id should be of category datatype, 

rest of the coulmn's datatype is correct

In [89]:
customer_transaction_data['date'] = pd.to_datetime(customer_transaction_data['date'])
customer_transaction_data['customer_id'] = customer_transaction_data['customer_id'].astype('category')
customer_transaction_data['item_id'] = customer_transaction_data['item_id'].astype('category')

In [90]:
customer_transaction_data.dtypes

date               datetime64[ns]
customer_id              category
item_id                  category
quantity                    int64
selling_price             float64
other_discount            float64
coupon_discount           float64
dtype: object

In [91]:
#checking for duplicate rows
customer_transaction_data.duplicated().sum()

2916

In [92]:
#printing duplicate rows along with all of the occurences
duplicate_rows = customer_transaction_data[customer_transaction_data.duplicated(keep=False)]
duplicate_rows

date customer_id item_id  quantity  selling_price  \
5463    2012-01-19         801   32650         3         213.72   
5465    2012-01-19         801   32650         3         213.72   
5491    2012-01-19         814   25251         2         163.14   
5493    2012-01-19         814   57670         2          21.37   
5497    2012-01-19         814   25251         2         163.14   
...            ...         ...     ...       ...            ...   
1323178 2013-07-03        1485   49741         1         142.12   
1323390 2013-07-03        1558   64186         2          71.24   
1323391 2013-07-03        1558   64186         2          71.24   
1323412 2013-07-03         998   18118         2         152.45   
1323413 2013-07-03         998   18118         2         152.45   

         other_discount  coupon_discount  
5463            -159.22             0.00  
5465            -159.22             0.00  
5491               0.00             0.00  
5493               0.00             0.00  
5497               0.00             0.00  
...                 ...              ...  
1323178          -35.62           -71.24  
1323390          -20.66           -71.24  
1323391          -20.66           -71.24  
1323412            0.00           -17.81  
1323413            0.00           -17.81  

[5747 rows x 7 columns]

Thus, we can observe from above data, that there are 2916 duplicate rows present in customer_transaction_data.

In [94]:
#dropping the duplicate rows
customer_transaction_data.drop_duplicates(inplace=True)

In [95]:
customer_transaction_data.duplicated().sum()

0

In [96]:
#get a summary of numerical features of the dataset
np.round(customer_transaction_data.describe(include=["int64", "float64"]), 2)

quantity  selling_price  other_discount  coupon_discount
count  1321650.00     1321650.00      1321650.00       1321650.00
mean       130.89         114.57          -17.74            -0.57
std       1312.46         152.73           37.82             7.01
min          1.00           0.36        -3120.31         -1992.23
25%          1.00          49.51          -23.15             0.00
50%          1.00          78.01           -1.78             0.00
75%          1.00         124.31            0.00             0.00
max      89638.00       17809.64            0.00             0.00

In [97]:
# statistical summary of categorical columns
customer_transaction_data.describe(include="category")

customer_id  item_id
count       1321650  1321650
unique         1582    74063
top            1555    49009
freq           4508    13536

From the above data, we observe:
1. For most of the transactions, quantity is 1 , but few transactions have taken place in large quantity , suggesting bulk purchase.
2. Most of the transactions have taken place without applying any coupon_discount.
3. For the other_discount, median is -1.78 , also as 75 percentile is 0, shows most of the transactions have taken place without applying any other_discount.
4. There are 1582 unique customer ids who looks for buying 74063 unique items.

Analysis of item_data

In [100]:
#considering item_data dataframe
item_data.sample(10)

item_id  brand   brand_type                category
5606      5607    848  Established          Pharmaceutical
29824    29825     46  Established          Pharmaceutical
671        672    946  Established                 Grocery
10975    10976   4807  Established          Pharmaceutical
27448    27449   2358  Established                 Alcohol
63522    63523   5302  Established        Natural Products
27710    27711   1124  Established                 Grocery
16691    16692   4972  Established  Dairy, Juices & Snacks
42807    42808     56        Local                 Grocery
27648    27649    676  Established                 Grocery

In [101]:
item_data.shape

(74066, 4)

In [102]:
item_data.dtypes

item_id        int64
brand          int64
brand_type    object
category      object
dtype: object

item_id should be of type category,
brand represents unique id for item brand, it should be of type category
brand_type and category should be of type category

In [104]:
item_data['item_id'] = item_data['item_id'].astype('category')
item_data['brand'] = item_data['brand'].astype('category')
item_data['brand_type'] = item_data['brand_type'].astype('category')
item_data['category'] = item_data['category'].astype('category')


In [105]:
item_data.dtypes

item_id       category
brand         category
brand_type    category
category      category
dtype: object

In [106]:
#checking for duplicate rows
item_data.duplicated().sum()

0

In [107]:
# statistical summary of categorical columns
item_data.describe(include="category")

item_id  brand   brand_type category
count     74066  74066        74066    74066
unique    74066   5528            2       19
top           1     56  Established  Grocery
freq          1  10480        62842    32448

From the above data, we observe:
1. There are no null values in any column in item_data.
2. There are 74066 unique ids corresponding to each item_id.
3. brand_type contains 2 different categories but people purchase more from an established brand.
4. From 19 diferent categories of items, grocery is most common.

Merging dataframes

Merging item_data with coupon_item_mapping

In [111]:
#creating sets for item_id for both the item_data and coupon_item_mapping dataframe
item_data_set = set(item_data.item_id.values)
coupon_item_mapping_set = set(coupon_item_mapping.item_id.values)

In [112]:
#finding the coupon_id not in item_data dataframe
coupon_item_mapping_set - coupon_item_mapping_set.intersection(item_data_set)

set()

All the item_id which are present in coupon_item_mapping and item_data

In [114]:
#merging item_data with coupon_item_mapping
item_coupon_merged = pd.merge(coupon_item_mapping,item_data, on='item_id', how='left')
item_coupon_merged

coupon_id  item_id brand   brand_type        category
0           105       37    56        Local         Grocery
1           107       75    56        Local         Grocery
2           494       76   209  Established         Grocery
3           522       77   278  Established         Grocery
4           518       77   278  Established         Grocery
...         ...      ...   ...          ...             ...
92658        32    69268   686  Established  Pharmaceutical
92659        32    68502  1009  Established  Pharmaceutical
92660        32    68612  1104  Established  Pharmaceutical
92661        33    71390  4196  Established            Meat
92662        31    70030  4477  Established         Grocery

[92663 rows x 5 columns]

Merging item_coupon_merged with train and test data

In [116]:
train = pd.merge(train,item_coupon_merged, on='coupon_id', how='left')
test = pd.merge(test,item_coupon_merged, on='coupon_id', how='left')

In [117]:
train

id  campaign_id  coupon_id customer_id redemption_status  \
0             1           13         27        1053                 0   
1             1           13         27        1053                 0   
2             1           13         27        1053                 0   
3             1           13         27        1053                 0   
4             1           13         27        1053                 0   
...         ...          ...        ...         ...               ...   
6420689  128595           13        681         623                 0   
6420690  128595           13        681         623                 0   
6420691  128595           13        681         623                 0   
6420692  128595           13        681         623                 0   
6420693  128595           13        681         623                 0   

        campaign_type start_date   end_date  campaign_duration  item_id brand  \
0                   X 2013-05-19 2013-07-05                 47    24775  1636   
1                   X 2013-05-19 2013-07-05                 47    14958  1636   
2                   X 2013-05-19 2013-07-05                 47    40431  1636   
3                   X 2013-05-19 2013-07-05                 47    20749  1636   
4                   X 2013-05-19 2013-07-05                 47    56860  1636   
...               ...        ...        ...                ...      ...   ...   
6420689             X 2013-05-19 2013-07-05                 47    16981   686   
6420690             X 2013-05-19 2013-07-05                 47    44676   686   
6420691             X 2013-05-19 2013-07-05                 47    38641   686   
6420692             X 2013-05-19 2013-07-05                 47    34285   686   
6420693             X 2013-05-19 2013-07-05                 47      990   686   

          brand_type category  
0        Established  Grocery  
1        Established  Grocery  
2        Established  Grocery  
3        Established  Grocery  
4        Established  Grocery  
...              ...      ...  
6420689  Established  Grocery  
6420690  Established  Grocery  
6420691  Established  Grocery  
6420692  Established  Grocery  
6420693  Established  Grocery  

[6420694 rows x 13 columns]

In [118]:
test

id  campaign_id  coupon_id customer_id campaign_type start_date  \
0             3           22        869         967             X 2013-09-16   
1             3           22        869         967             X 2013-09-16   
2             3           22        869         967             X 2013-09-16   
3             3           22        869         967             X 2013-09-16   
4             3           22        869         967             X 2013-09-16   
...         ...          ...        ...         ...           ...        ...   
5421796  128594           18        988         851             X 2013-08-10   
5421797  128594           18        988         851             X 2013-08-10   
5421798  128594           18        988         851             X 2013-08-10   
5421799  128594           18        988         851             X 2013-08-10   
5421800  128594           18        988         851             X 2013-08-10   

          end_date  campaign_duration  item_id brand   brand_type  \
0       2013-10-18                 32     1033  1075  Established   
1       2013-10-18                 32     1067  1075  Established   
2       2013-10-18                 32     1084  1075  Established   
3       2013-10-18                 32     1124  1075  Established   
4       2013-10-18                 32     1125  1075  Established   
...            ...                ...      ...   ...          ...   
5421796 2013-10-04                 55    53817  1342  Established   
5421797 2013-10-04                 55    55314  1342  Established   
5421798 2013-10-04                 55    20798  1342  Established   
5421799 2013-10-04                 55    62761  1342  Established   
5421800 2013-10-04                 55    62835  1342  Established   

               category  
0               Grocery  
1               Grocery  
2               Grocery  
3               Grocery  
4               Grocery  
...                 ...  
5421796  Pharmaceutical  
5421797  Pharmaceutical  
5421798  Pharmaceutical  
5421799  Pharmaceutical  
5421800  Pharmaceutical  

[5421801 rows x 12 columns]

In [119]:
#merging customer_item_demographics with train and test data
train = pd.merge(train,customer_demographics, on='customer_id', how='left')
test = pd.merge(test,customer_demographics, on='customer_id', how='left')

In [120]:
train.head()

id  campaign_id  coupon_id  customer_id redemption_status campaign_type  \
0  1           13         27         1053                 0             X   
1  1           13         27         1053                 0             X   
2  1           13         27         1053                 0             X   
3  1           13         27         1053                 0             X   
4  1           13         27         1053                 0             X   

  start_date   end_date  campaign_duration  item_id brand   brand_type  \
0 2013-05-19 2013-07-05                 47    24775  1636  Established   
1 2013-05-19 2013-07-05                 47    14958  1636  Established   
2 2013-05-19 2013-07-05                 47    40431  1636  Established   
3 2013-05-19 2013-07-05                 47    20749  1636  Established   
4 2013-05-19 2013-07-05                 47    56860  1636  Established   

  category age_range marital_status rented family_size no_of_children  \
0  Grocery     46-55            NaN      0           1            NaN   
1  Grocery     46-55            NaN      0           1            NaN   
2  Grocery     46-55            NaN      0           1            NaN   
3  Grocery     46-55            NaN      0           1            NaN   
4  Grocery     46-55            NaN      0           1            NaN   

  income_bracket  
0              5  
1              5  
2              5  
3              5  
4              5

In [121]:
train.shape

(6420694, 19)

In [211]:
test.shape

(5421801, 18)